### DISCUSSION TITLE
Laravel: Spatie Activity Log / Show causer in timeline widget

### DISCUSSION THREAD

**Discussion by FrazeColder** [filamentphp/filament#14067] [Help]
### Package

Form builder

### Package Version

v3.2.1

### How can we help you?

I'm using the Activitylog Pro package ([link][1]) and running into an issue with displaying log details in my FilamentPHP OrderResource.

**Context**
- I have two models: `Order` and `Disapprove`.
- In my `OrderResource`, I display a timeline of activities.
- When a customer disapproves an order (which can happen multiple times for custom orders), a log entry is created.
- My goal is to show a button which opens a modal with the disapprove reason within this timeline to display the disapprove reason in a modal.

This is my current widget:
```
Timeline::make('activity')
    ->label('History')
    ->itemActions('disapprove', [
        Forms\Components\Actions\Action::make('disapprove_reason')
            ->label('Disapprove Reason')
            ->modalHeading('Disapprove reason')
            ->modalDescription(fn ($record) => $record->id)
            ->modalSubmitAction(false)
            ->modalCancelActionLabel('Close')
            ->modalCancelAction(),
    ]);
```

**Problem**

The `$record` variable in the modal always refers to the `Order` model. I need it to reference the specific Disapprove model associated with the log entry.

I store the Disapprove model as the "causer" of the log entry, which might not be the best practice, but it’s what I’m working with. How can I modify the code to achieve this and retrieve the "causer" for each log entry to pass the data to the modal?

As already mentioned, it does not make sense to just retrieve the most recent log entry. This way all buttons/modals would display the most recent disapproval reason and not the one I wanted to open.

Thanks in advance for your help!

  [1]: https://filamentphp.com/plugins/ralphjsmit-activitylog-pro

**Comment by rvzug** [comment]
If your ```$record``` is always a ```Order```, and Order has a relation ```->causer()```, than you could use ```fn ($record) => $record->causer->some_attribute``` on places where you can use a callback.

Or do you mean that the $order->causer() relation returns multiple ```Disapprove``` models?

**Reply by FrazeColder** [reply]
This does not work because `->causer()` is a relationship of the `ActivityLog` model by spatie/activity-log and not the `Order` model. That is the problem I am facing here...

**Reply by rvzug** [reply]
So if I understand your question correctly, you need to define a (inverse) [polymorphic relationship](https://laravel.com/docs/11.x/eloquent-relationships#one-to-many-polymorphic-relations) between your ```Order``` model, and the ```Activity``` model from the Spatie library. So, if you haven't yet, add:

```php
public function activities()
{
    return $this->morphMany(\Spatie\Activitylog\Models\Activity::class, 'subject');
}
```
to your Order model, to create an inverse on [the polymorphic relationship 'subject'](https://github.com/spatie/laravel-activitylog/blob/main/src/Models/Activity.php#L62) from Activity.

And than in your modal you could reach the information based on ```$record->activities...```:
```fn (Order $record) => $record->activities->latest()->description``` would give you the description of the latest activity.
```fn (Order $record) => $record->activities->latest()->causer->name``` would give you the name of the causer of the latest activity.

I think the current ```Action::modalDescription()``` method does not have enough space to display multiple activities. Maybe you could look to [passing a view to the model](https://filamentphp.com/docs/3.x/actions/modals#custom-modal-content) so you can loop trough all the ```$record->activities```

**Reply by FrazeColder** [reply]
This is exactly the point where I am currently. As you mentioned, the relationship is polymorphic.

However, I don't want to get only the latest log entry of my disapproval. I want to have the specific disapproval for each entry.

Look here, I have multiple disapproval and each button should map to the correct one:

<img width="899" alt="grafik" src="https://github.com/user-attachments/assets/933960a7-604f-4a91-9986-1cb0385e7379">
